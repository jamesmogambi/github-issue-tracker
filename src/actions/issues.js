require("isomorphic-fetch");

// SET_REPOSITORY_ISSUES
export const setRepositoryIssues = (payload) => ({
  type: "SET_REPOSITORY_ISSUES",
  payload,
});
// CLEAR_ISSUES
export const clearIssues = () => ({
  type: "CLEAR_ISSUES",
});
// ERROR
export const error = (payload) => ({
  type: "ERROR",
  payload,
});

// Get Github Repo Issues
export const getRepositoryIssues = (params) => {
  return (dispatch) => {
    dispatch(clearIssues());
    fetch("https://api.github.com/graphql", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        authorization: `bearer ${process.env.GITHUB_ACCESS_TOKEN}`,
      },

      body: JSON.stringify({
        query: `
                  query($repoOwner: String!, $repoName: String!) {
                  repository(name: $repoName, owner: $repoOwner) {
                    issues(first: 100) {
                      edges {
                        node {
                           id
                           number
                           state
                           createdAt
                           title
                           url
                           author{
                            avatarUrl
                            login
                          }
                           bodyHTML
                        }
                      }
                    }
                  }
                }
                 `,
        variables: {
          repoOwner: params.repoOwner,
          repoName: params.repoName,
        },
      }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (res.data.repository) {
          dispatch(
            setRepositoryIssues({
              repositoryissues: res.data.repository.issues.edges,
              username: params.repoOwner,
              repositoryname: params.repoName,
            })
          );
        }
      })
      .catch((err) => {
        dispatch(error());
        console.error(err);
      });
  };
};
