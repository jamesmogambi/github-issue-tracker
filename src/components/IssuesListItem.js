import React from "react";
import { Link } from "react-router-dom";

const IssuesListItem = ({ id, title, author, createdAt }) => (
  <Link className="list-item" to={`/issueDetails/${id}`}>
    <div>
      <img src={author.avatarUrl} alt="Avatar" className="avatar" />
      <div className="list-item__sub-title">{title}</div>
    </div>
    <span className="list-item__sub-title ">opened at {createdAt}</span>
  </Link>
);

export default IssuesListItem;
