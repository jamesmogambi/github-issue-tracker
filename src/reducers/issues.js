// Issues Reducer

const initialState = {
  username: null,
  repositoryname: null,
  repositoryissues: [],
  error: null,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case "SET_REPOSITORY_ISSUES":
      const { username, repositoryname, repositoryissues } = action.payload;
      return {
        ...state,
        username,
        repositoryname,
        repositoryissues,
      };

    case "CLEAR_ISSUES":
      return {
        ...state,
        username: null,
        repositoryname: null,
        repositoryissues: [],
      };
    default:
      return state;
  }
};
