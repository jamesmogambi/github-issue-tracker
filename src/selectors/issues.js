// Get visible issues
export default (issues, { text }) => {
  return issues.filter(({ node }) => {
    const textMatch = node.title.toLowerCase().includes(text.toLowerCase());
    return textMatch;
  });
};
