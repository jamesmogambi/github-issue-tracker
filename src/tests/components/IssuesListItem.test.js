import React from "react";
import { shallow } from "enzyme";
import issues from "../fixtures/issues";
import IssuesListItem from "../../components/IssuesListItem";

test("should render IssuesListItem correctly", () => {
  const wrapper = shallow(<IssuesListItem {...issues[0].node} />);
  expect(wrapper).toMatchSnapshot();
});
