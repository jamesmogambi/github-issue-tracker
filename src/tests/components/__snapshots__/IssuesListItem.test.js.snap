// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IssuesListItem correctly 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/issueDetails/MDU6SXNzdWU3Njc4NjI5Nzk="
>
  <div>
    <img
      alt="Avatar"
      className="avatar"
      src="https://avatars.githubusercontent.com/u/31744209?u=a6d48cf8f85179ba4db1b3268e8624fd21ddc2e5&v=4"
    />
    <div
      className="list-item__sub-title"
    >
      screenshots
    </div>
  </div>
  <span
    className="list-item__sub-title "
  >
    opened at 
    2020-12-15T17:20:41Z
  </span>
</Link>
`;
